File: Loenn/entities/decalEffectTarget.lua
local drawableSprite = require("structs.drawable_sprite")

local decalEffectTarget = {}

--decalEffectTarget.justification = { 0, 0 }

decalEffectTarget.name = "PuzzleIslandHelper/DecalEffectTarget"
function decalEffectTarget.depth(room, entity)
    return entity.depth or 0
end
decalEffectTarget.placements = {
    name = "Decal Effect Target",
    data = {
        fps = 12.0,
        decalPath = "1-forsakencity/flag",
        depth = 1,
        groupId = "decalEffectGroup_1",
        scaleX = 1,
        scaleY = 1,
        rotation = 0
    }
}
function decalEffectTarget.sprite(room, entity)
    local path;
    if drawableSprite.fromTexture("decals/" .. entity.decalPath .. "00") ~= nil then
        path = "decals/" .. entity.decalPath .. "00"
    else
        path = "decals/" .. entity.decalPath
    end
    local sprite = drawableSprite.fromTexture(path, entity)
        sprite:setScale(entity.scaleX or 1, entity.scaleY or 1)
        sprite:setJustification(0.5, 0.5)
        sprite.rotation = math.rad(entity.rotation or 0)
    return sprite
end

return decalEffectTarget

File: Entities/DecalEffectTarget.cs
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

// PuzzleIslandHelper.DecalEffectTarget
namespace Celeste.Mod.PuzzleIslandHelper.Entities
{
    [CustomEntity("PuzzleIslandHelper/DecalEffectTarget")]
    [Tracked]
    public class DecalEffectTarget : Entity
    {
        public Sprite sprite;
        public string ID;
        private string id;

        public DecalEffectTarget(EntityData data, Vector2 offset)
        : base(data.Position + offset)
        {
            Depth = data.Int("depth", 2);
            float delay = 1f / (data.Float("fps") / 2f);
            sprite = new Sprite(GFX.Game, "decals/");
            string path = data.Attr("decalPath");
            if (path.Contains("decals/"))
            {
                path.Replace("decals/", "");
            }
            sprite.AddLoop("idle", data.Attr("decalPath"), 0.1f);
            id = data.Attr("groupId");
            Add(sprite);
            sprite.Scale = new Vector2(data.Float("scaleX", 1), data.Float("scaleY", 1));
            sprite.Rotation = data.Float("rotation").ToRad();
            sprite.CenterOrigin();
            sprite.Position += new Vector2(sprite.Width / 2, sprite.Height / 2);
            sprite.Visible = false;
            Collider = new Hitbox(sprite.Width, sprite.Height, -sprite.Width / 2, -sprite.Height / 2);
            Tag |= Tags.TransitionUpdate;
        }
        public override void Added(Scene scene)
        {
            base.Added(scene);
            ID = (scene as Level).Session.Level + "_decaltargetgroup_" + id;
            sprite.Play("idle");
        }
    }
}
